lib/builders.ncl

```diff
@@ -84,4 +93,75 @@ in
         structured_env.buildInputs = packages,
       }
         | (NickelPkg & { packages | { _ : Derivation } }),
+
+  ShellApplication
+    | doc m%"
+    "%
+    =
+      {
+        name,
+        version | default = "0.0.0",
+        text | NullOr NixString | MutExclusiveWith file "text" "file" | default = null,
+        file | NullOr String | default = null,
```

Comment:
Actually, another more idiomatic solution might be

```nickel
{
  content.text | optional | NixString,
  content.file | NixString | default = lib.to_file content.text,
}
```

(where `lib.to_file` has to be implemented, but it's something quite useful any way)

Reply:
> `TaggedUnion`

Is it builtin or implemented somewhere? I'm not sure how to implement it preserving record contract semantics for LSPs...

> `lib.to_file`

I actually wanted to avoid creation of almost duplicate paths in /nix/store if possible, so if the text is available in Nickel, I'd prefer to write it to .drv rather than into separate file. Although I might be convinced otherwise...

Reply:
What's the difference between having it in the drv and in a separate file?

Reply:
> Is it builtin or implemented somewhere? I'm not sure how to implement it preserving record contract semantics for LSPs...

It's indeed not implemented. But it could reuse the same logic as `MutExclusiveWith`, just with one layer of nesting (not pretty, but that would work)

Reply:
> we won't be able to enforce setting either text or file fields with contract.

No, but that's actually probably required for overriding to work properly. We probably want something like the above to work:

```nickel
let myApp = {
  name = "foo",
  content.file = "foo.sh",
} in
myApp & {
  content.text = "echo Overriden"
}
```

(ideally we would have that if `content.file` is defined, then `content.text` has to be `force`-ed to be overridden, but I don't think that's possible)